### PR TITLE
Optimizing the performance of fused_layer_norm and top_p_sampling operators

### PR DESCRIPTION
**Pull request by yuanlehome**
### PR Category

Inference

### PR Types

Performance

### Description

pcard-71500
Optimizing the performance of fused_layer_norm and top_p_sampling operators.

### CODE REVIEW

File: paddle/phi/kernels/gpu/top_p_sampling_kernel.cu [hunk 1]
```diff
@@ -35,6 +35,12 @@ namespace cub = hipcub;
 #include "paddle/phi/kernels/funcs/top_k_function_cuda.h"
 #include "paddle/phi/kernels/primitive/functor_primitives.h"
 
+#ifdef PADDLE_WITH_HIP
+#define GPU(str) hip##str
+#else
+#define GPU(str) cu##str
```
Comment (YanhuiDua) [inline paddle/phi/kernels/gpu/top_p_sampling_kernel.cu:41]: #define GPU(str) cuda##str

File: paddle/phi/kernels/gpu/top_p_sampling_kernel.cu [hunk 2]
```diff
@@ -35,6 +35,12 @@ namespace cub = hipcub;
 #include "paddle/phi/kernels/funcs/top_k_function_cuda.h"
 #include "paddle/phi/kernels/primitive/functor_primitives.h"
 
+#ifdef PADDLE_WITH_HIP
+#define GPU(str) hip##str
```
Comment (YanhuiDua) [inline paddle/phi/kernels/gpu/top_p_sampling_kernel.cu:39]: #include <hiprand_kernel.h>

File: paddle/phi/kernels/gpu/top_p_sampling_kernel.cu [hunk 3]
```diff
@@ -881,9 +730,11 @@ __global__ void topp_sampling(T* sorted_probs,
...
         (thread_offset >= p_t && thread_offset - thread_count < p_t)) {
 #ifdef PADDLE_WITH_HIP
-      float random_ratio = exponential_transform(hiprand_uniform(&state), 1.0f);
+      float random_ratio =
+          exponential_transform(hiprand_uniform(states + bid), 1.0f);
 #else
-      float random_ratio = exponential_transform(curand_uniform(&state), 1.0f);
+      float random_ratio =
+          exponential_transform(curand_uniform(states + bid), 1.0f);
 #endif
```
Comment (YanhuiDua) [inline paddle/phi/kernels/gpu/top_p_sampling_kernel.cu]: 这里也可以用GPU(rand_uniform)包起来

File: paddle/phi/kernels/gpu/top_p_sampling_kernel.cu [hunk 4]
```diff
@@ -1137,114 +977,90 @@ void DispatchTopPSampling(const Context& dev_ctx,
...
+__global__ void setup_kernel(GPU(randState_t) * state,
+                             const uint64_t seed,
+                             const uint64_t offset,
+                             const int bs,
+                             const bool need_batch_random) {
+  int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  for (int i = idx; i < bs; i += gridDim.x * blockDim.x) {
+    if (need_batch_random) {
+#ifdef PADDLE_WITH_HIP
+      hiprand_init(seed, i, offset, &state[i]);
```
Comment (YanhuiDua) [inline paddle/phi/kernels/gpu/top_p_sampling_kernel.cu]: 函数也可以用宏定义：GPU(rand_init)

Comment (yuanlehome) [reply paddle/phi/kernels/gpu/top_p_sampling_kernel.cu]: done

File: paddle/phi/kernels/gpu/top_p_sampling_kernel.cu (hunk 3 above)
Comment (yuanlehome) [reply paddle/phi/kernels/gpu/top_p_sampling_kernel.cu]: done

**Review by carryyu — APPROVED** [review]

File: paddle/phi/kernels/gpu/top_p_sampling_kernel.cu [hunk 5]
```diff
@@ -633,133 +571,61 @@ template <typename T, typename Context, int TopKMaxLength, int TopPBeamTopK>
...
-                                                     seed_offset,
-                                                     count_iter,
-                                                     count_iter_begin,
-                                                     k,
-                                                     need_batch_random));
-        default:
-          PD_THROW(
-              "the input data shape has error in the topp_beam_topk kernel.");
-      }
+  if (mode == "truncate") {
```
Comment (YanhuiDua) [inline paddle/phi/kernels/gpu/top_p_sampling_kernel.cu:588]: 需要对应修改search.py的python接口的默认mode 为“truncate”

**Pull request merged**